barretenberg/cpp/src/barretenberg/relations/logderiv_lookup_relation.hpp

```diff
@@ -157,16 +158,25 @@ template <typename FF_> class LogDerivLookupRelationImpl {
     {
         auto& inverse_polynomial = get_inverse_polynomial(polynomials);
 
-        for (size_t i = 0; i < circuit_size; ++i) {
-            // We only compute the inverse if this row contains a lookup gate or data that has been looked up
-            if (polynomials.q_lookup.get(i) == 1 || polynomials.lookup_read_tags.get(i) == 1) {
-                // TODO(https://github.com/AztecProtocol/barretenberg/issues/940): avoid get_row if possible.
-                auto row = polynomials.get_row(i); // Note: this is a copy. use sparingly!
-                auto value = compute_read_term<FF, 0>(row, relation_parameters) *
-                             compute_write_term<FF, 0>(row, relation_parameters);
-                inverse_polynomial.at(i) = value;
+        size_t min_iterations_per_thread = 1 << 6; // min number of iterations for which we'll spin up a unique thread
+        size_t num_threads = bb::calculate_num_threads_pow2(circuit_size, min_iterations_per_thread);
+        size_t iterations_per_thread = circuit_size / num_threads; // actual iterations per thread
+
+        parallel_for(num_threads, [&](size_t thread_idx) {
+            size_t start = thread_idx * iterations_per_thread;
+            size_t end = (thread_idx + 1) * iterations_per_thread;
+            for (size_t i = start; i < end; ++i) {
+                // We only compute the inverse if this row contains a lookup gate or data that has been looked up
+                if (polynomials.q_lookup.get(i) == 1 || polynomials.lookup_read_tags.get(i) == 1) {
+                    // TODO(https://github.com/AztecProtocol/barretenberg/issues/940): avoid get_row if possible.
+                    auto row = polynomials.get_row(i); // Note: this is a copy. use sparingly!
```

Comment:
Just a comment because I see a lot of microoptimisation going on: have you guys profiled where the time is going?

IIRC when I did some profiling for the AVM (which luckily does not use this relation, since we already parallelize) a lot of time was spent in
- Using `get_row` which (according to some comments) "was only meant for debugging"
- (related to the above) Visiting all rows of the trace and finding out if you actually need to compute the inverse or not.

If this is still true, you'd benefit far more if you could save which rows need an inverse, and only act on those (therefore O(needs_inverse) vs O(rows)). I think this is true for the AVM (where the cost of get_row is really big) and I have some ideas on how to implement this. However, for general bb, the problem is that the whole proving system is stateless in terms of relations. That is, everything uses `Relation::method`, and not `relation_instance.method`.

Reply:
I actually looked into the issue here https://github.com/AztecProtocol/barretenberg/issues/940 and we indeed only use `get_row` for those that need an inverse (achieved by checking `q_lookup` and `lookup_read_target` polynomial). I'm optimising with the scope of reducing the impact of using a larger trace structure when keeping the same circuit content (so introducing more sparsity in the trace). I did indeed profiled `get_row` and the time spent doing it is not getting worse because we avoid calling it when not needed. We still are spending around 3.5s calling `get_row` but as Luke noted in the issue moving to a different pattern requires more involved work and we want to eliminate the existing low hanging fruits first.